### PR TITLE
Stratify apiserver cert generation

### PR DESCRIPTION
**Pull request by sttts**
- move self-signed cert generation to `SecureServingOptions.MaybeDefaultWithSelfSignedCerts`
- make cert generation only depend on `ServerRunOptions`, not on an unfinished `Config` (this breaks the chicken-egg problem of a finished config in https://github.com/kubernetes/kubernetes/pull/35387#pullrequestreview-5368176)
- move loopback client config code into `config_selfclient.go`

Replaces https://github.com/kubernetes/kubernetes/pull/35387#event-833649341 by getting rid of duplicated `Complete`.

### CODE REVIEW

**Comment by k8s-oncall** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37830)

**Comment by deads2k** [comment]
I liked the old pull name better :)

File: pkg/genericapiserver/config.go [hunk 1]
```diff
@@ -235,45 +224,98 @@ func NewConfig() *Config {
...
 		ClientCA: secureServing.ClientCA,
 	}
-	if secureServing.ServerCert.CertKey.CertFile == "" && secureServing.ServerCert.CertKey.KeyFile == "" {
-		secureServingInfo.ServerCert.Generate = true
-		secureServingInfo.ServerCert.CertFile = path.Join(secureServing.ServerCert.CertDirectory, secureServing.ServerCert.PairName+".crt")
-		secureServingInfo.ServerCert.KeyFile = path.Join(secureServing.ServerCert.CertDirectory, secureServing.ServerCert.PairName+".key")
+
+	serverCertFile, serverKeyFile := secureServing.ServerCert.CertKey.CertFile, secureServing.ServerCert.CertKey.KeyFile
+
+	// possibly self-signed cert and key
```
Comment (sttts) [inline pkg/genericapiserver/config.go]: we can split those two steps for that reason. Had the same idea, but didn't implement it yet. Maybe in a follow-up?

Comment (deads2k) [inline pkg/genericapiserver/config.go]: > we can split those two steps for that reason. Had the same idea, but didn't implement it yet. Maybe in a follow-up?

Most of this pull will remain right? It just promotes the code out of here.  I'm not really a fan of having this method behave this way.

**Comment by sttts** [comment]
@k8s-bot test this issue

**Comment by sttts** [comment]
@deads2k ptal at the 2nd commit. Looks like the e2e tests do not prepare certs with "localhost". So we fall back if the insecure port is open anyway, but stay secure if the insecure port is disabled.

**Comment by deads2k** [comment]
> @deads2k ptal at the 2nd commit. Looks like the e2e tests do not prepare certs with "localhost". So we fall back if the insecure port is open anyway, but stay secure if the insecure port is disabled.

I'm ok with it for now, but we should try to get the e2e tests updated.  @kubernetes/test-infra-maintainers

**Comment by ixdy** [comment]
@kubernetes/sig-testing ^ previous commennt by @deads2k

**Comment by sttts** [comment]
@k8s-bot gci gke e2e test this

**Comment by sttts** [comment]
@k8s-bot gci gke e2e test this

**Comment by sttts** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by sttts** [comment]
@deads2k may I read your last comment as lgtm?

File: federation/cmd/federation-apiserver/app/server.go [hunk 2]
```diff
@@ -73,16 +73,19 @@ func Run(s *options.ServerRunOptions) error {
...
 
-	genericapiserver.DefaultAndValidateRunOptions(s.GenericServerRunOptions)
-	genericConfig := genericapiserver.NewConfig(). // create the new config
-							ApplyOptions(s.GenericServerRunOptions). // apply the options selected
-							ApplySecureServingOptions(s.SecureServing).
-							ApplyInsecureServingOptions(s.InsecureServing).
-							ApplyAuthenticationOptions(s.Authentication).
-							ApplyRBACSuperUser(s.Authorization.RBACSuperUser)
+	if err := s.SecureServing.MaybeDefaultWithSelfSignedCerts(s.GenericServerRunOptions.AdvertiseAddress.String()); err != nil {
+		glog.Fatalf("Error creating self-signed certificates: %v", err)
```
Comment (deads2k) [inline federation/cmd/federation-apiserver/app/server.go]: return error

File: cmd/kube-apiserver/app/server.go [hunk 3]
```diff
@@ -85,20 +85,25 @@ func Run(s *options.ServerRunOptions) error {
...
-
 	serviceIPRange, apiServerServiceIP, err := master.DefaultServiceIPRange(s.GenericServerRunOptions.ServiceClusterIPRange)
 	if err != nil {
 		glog.Fatalf("Error determining service IP ranges: %v", err)
 	}
-	if err := genericConfig.MaybeGenerateServingCerts(apiServerServiceIP); err != nil {
-		glog.Fatalf("Failed to generate service certificate: %v", err)
+
+	if err := s.SecureServing.MaybeDefaultWithSelfSignedCerts(s.GenericServerRunOptions.AdvertiseAddress.String(), apiServerServiceIP); err != nil {
+		glog.Fatalf("Error creating self-signed certificates: %v", err)
```
Comment (deads2k) [inline cmd/kube-apiserver/app/server.go]: return error

File: pkg/genericapiserver/options/serving.go [hunk 4]
```diff
@@ -59,6 +55,8 @@ type CertKey struct {
 type GeneratableKeyCert struct {
 	CertKey CertKey
 
+	// CaCertFile is an optional file containing the certificate chain for CertKey.CertFile
+	CaCertFile string
```
Comment (deads2k) [inline pkg/genericapiserver/options/serving.go]: CACertFile

**Comment by deads2k** [comment]
> @deads2k may I read your last comment as lgtm?

Very minor comments.  I don't need another look.  Fix up, squash in comments, and tag.

File: cmd/kube-apiserver/app/server.go (hunk 3 above)
Comment (sttts) [inline cmd/kube-apiserver/app/server.go]: this file is full of glog.Fatalf's

Comment (deads2k) [inline cmd/kube-apiserver/app/server.go]: > this file is full of glog.Fatalf's

ok, we can tidy later.

**Pull request merged**